### PR TITLE
Make any and all on ByteTensor behave like sum/prod.

### PR DESCRIPTION
**Pull request by sighingnow**
Make `ByteTensor`'s  methods `any` and `all` can accept `dim` and `keepdim` parameters, behave like `sum/prod`.

I add two macro `TENSOR_IMPLEMENT_ACC` and `TENSOR_IMPLEMENT_ACCALL` to make the code less duplicate.

This feature should fix issue #4313

### CODE REVIEW

File: aten/src/TH/generic/THTensorMath.c [hunk 1]
```diff
@@ -1870,78 +1867,44 @@ void THTensor_(min)(THTensor *values_, THLongTensor *indices_, THTensor *t, int
...
-
-    TH_TENSOR_APPLY2(real, temp_, real, t, *temp__data = *temp__data * *t_data;);
-    THTensor_(free)(temp_);
-  }
-
-  if (!keepdim) {
-    THTensor_(squeeze1d)(r_, r_, dimension);
-  }
-}
+#define TENSOR_IMPLEMENT_ACC(NAME, ACC_OP, INIT_VALUE)                                   \
```
Comment (apaszke) [inline aten/src/TH/generic/THTensorMath.c]: Can you add a comment that `ACC_OP` needs to be commutative?

File: aten/src/TH/generic/THTensorMath.c [hunk 2]
```diff
@@ -3040,16 +3003,11 @@ LAB_IMPLEMENT_BASIC_FUNCTION(abs,abs)
...
 #if defined(TH_REAL_IS_BYTE)
 
-#define TENSOR_IMPLEMENT_LOGICAL_SUM(NAME, OP, INIT_VALUE) \
-  int THTensor_(NAME)(THTensor *tensor) \
-  { \
-    int sum = INIT_VALUE;                               \
-    TH_TENSOR_APPLY(real, tensor, sum = sum OP *tensor_data;); \
-    return sum; \
-  }
+TENSOR_IMPLEMENT_ACCALL(logicalallall, int, &&, 1)
```
Comment (apaszke) [inline aten/src/TH/generic/THTensorMath.c]: Can we rename them to `logicalAndAll` and `logicalAnd` (for the one that reduces over a dimension)?

File: test/test_torch.py [hunk 3]
```diff
@@ -461,6 +461,22 @@ def test_all_any_empty(self):
         self.assertTrue(x.all())
         self.assertFalse(x.any())
 
+    @unittest.skipIf(not TEST_NUMPY, "Numpy not found")
+    def test_all_any_with_dim(self):
+        def test(x):
+            r1 = np.all(x, axis=1, keepdims=False).astype('uint8')
```
Comment (apaszke) [inline test/test_torch.py]: Can you do `x.float().sum() > 0` for OR, and `== x.dim(dim)` for AND? That would remove the numpy dependency

File: torch/lib/THD/master_worker/common/Functions.hpp [hunk 4]
```diff
@@ -208,7 +208,9 @@ enum Functions: std::uint16_t {
   tensorLogspace,
   tensorRand,
   tensorRandn,
+  tensorLogicalallall,
```
Comment (apaszke) [inline torch/lib/THD/master_worker/common/Functions.hpp]: The changes in this part of THD aren't necessary. We'll be refactoring it to automatically generate this code. Still you don't have to revert these changes, just keep that in mind for the future

File: torch/csrc/generic/methods/TensorCompare.cwrap [hunk 5]
```diff
@@ -737,24 +737,62 @@
...
+      arguments:
+        - arg: THTensor* result
+          output: True
+        - THTensor* self
+        - arg: int64_t dim
+          wrap_dim: self
+        - bool keepdim
+    - cname: logicalany
+      return: argument 0
+      before_call: maybeThrowBackCompatKeepdimWarn("any");
```
Comment (apaszke) [inline torch/csrc/generic/methods/TensorCompare.cwrap]: Warnings aren't necessary. These function didn't even have reducing overloads before we deprecated them, so we can assume people already follow the new guidelines

**Review by apaszke — COMMENTED** [review]
That looks good, thanks for the PR. I only have a few minor comments

**Comment by sighingnow** [comment]
Thanks for your review, @apaszke , new patch done.

File: test/test_torch.py (hunk 3 above)
Comment (sighingnow) [reply test/test_torch.py]: I have fixed the test using `torch.sum` and `torch.prod`, removing the numpy dependency.

File: torch/lib/THD/master_worker/common/Functions.hpp (hunk 4 above)
Comment (sighingnow) [reply torch/lib/THD/master_worker/common/Functions.hpp]: Thanks for your clarification. I also rename `logicalall` to `logicalAnd` and `logicalallall` to `logicalAndAll` for consistence.

**Review by gchanan — COMMENTED** [review]
looks good, thanks!  Some small suggestions below, and I'll fix the general ATen issue about bool returns.

**Comment by gchanan** [comment]
https://github.com/pytorch/pytorch/pull/4639 is in.

File: aten/src/ATen/Declarations.cwrap [hunk 6]
```diff
@@ -1113,25 +1113,53 @@
...
   backends:
     - CPU
     - CUDA
-  cname: logicalall
-  return: bool
-  arguments:
-    - THTensor* self
+  options:
+    - cname: logicalAndAll
+      return: bool
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply aten/src/ATen/Declarations.cwrap]: I will rebase to master.

File: aten/src/TH/generic/THTensorMath.c [hunk 7]
```diff
@@ -3040,16 +3008,11 @@ LAB_IMPLEMENT_BASIC_FUNCTION(abs,abs)
...
-    int sum = INIT_VALUE;                               \
-    TH_TENSOR_APPLY(real, tensor, sum = sum OP *tensor_data;); \
-    return sum; \
-  }
+TENSOR_IMPLEMENT_ACCALL(logicalAndAll, int, &&, 1)
+TENSOR_IMPLEMENT_ACCALL(logicalAnyAll, int, ||, 0)
 
-TENSOR_IMPLEMENT_LOGICAL_SUM(logicalall, &&, 1)
-TENSOR_IMPLEMENT_LOGICAL_SUM(logicalany, ||, 0)
+TENSOR_IMPLEMENT_ACC(logicalAnd, &&, 1)
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply aten/src/TH/generic/THTensorMath.c]: Is it ok to rename it as `logicaland` and `logicalandall` ?

[the remark this answers is not in the document]
Comment (sighingnow) [reply aten/src/TH/generic/THTensorMath.c]: Or as `logical_and` and `logical_and_all` ?

File: aten/src/TH/generic/THTensorMath.c [hunk 8]
```diff
@@ -734,19 +734,19 @@ real THTensor_(medianall)(THTensor *tensor)
...
-accreal THTensor_(sumall)(THTensor *tensor)
-{
-  accreal sum = 0;
-  TH_TENSOR_APPLY(real, tensor, sum += *tensor_data;);
-  return sum;
-}
+// Reduce all elements in the given tensor `t` with given `INIT_VALUE` and return
+// the result of type `RETTYPE`.
+// The accumulate operator `ACC_OP` must be commutative.
+#define TENSOR_IMPLEMENT_ACCALL(NAME, RETTYPE, ACC_OP, INIT_VALUE)        \
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply aten/src/TH/generic/THTensorMath.c]: Fixed.

File: aten/src/TH/generic/THTensorMath.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (sighingnow) [reply aten/src/TH/generic/THTensorMath.c]: I have also noticed than we use camel case in this file like `indexCopy`, `dataOffset` and `wrapLinearIndex`

**Comment by sighingnow** [comment]
Ping @gchanan

**Comment by ezyang** [comment]
This needs a rebase now; sorry about the delay :(

**Comment by sighingnow** [comment]
Thanks for picking up this. @ezyang I have rebased this PR to current master.

**Comment by sighingnow** [comment]
Fix compile error in THD module.

**Comment by sighingnow** [comment]
The failures wasn't caused by the code. Seems something bad happened to the CI server.

File: aten/src/TH/generic/THTensorMath.c [hunk 9]
```diff
@@ -3814,16 +3814,205 @@ LAB_IMPLEMENT_BASIC_FUNCTION(abs,abs)
...
+#else
+    serial_path = 1;
+#endif
+  if (serial_path) {
+    TH_TENSOR_APPLY(real, tensor, sum = sum || *tensor_data;);
+  }
+  return (bool)sum;
+}
+
+void THTensor_(logicalAnd)(THTensor *r_, THTensor *t, int dimension, int keepdim)
```
Comment (apaszke) [inline aten/src/TH/generic/THTensorMath.c]: This is code copied straight from `sum`, right (except for the binary ops)?

**Review by apaszke — APPROVED** [review]

File: aten/src/TH/generic/THTensorMath.c (hunk 9 above)
Comment (sighingnow) [reply aten/src/TH/generic/THTensorMath.c]: Yes. Previously (before the openmp patch for `add/prod`), I implemented the `and` and 'all' via macros, we only needed to write the function (as a macro definition) once then use macro to get `add/prod/all/any`. However after the openmp patch, the function itself use macros to detect if openmp exists, then it's impossible to definition to define the function as a macro definition (due to the need for nested macro definition, the `#ifdef` will need to occur inside a macro).

It's quite uncomfortable to copy-and-paste, but I think there's  no better choice :(

Comment (apaszke) [reply aten/src/TH/generic/THTensorMath.c]: I agree, but hopefully this code will go away soon with our upcoming refactors, so it's not a big deal.

**Comment by soumith** [comment]
```
15:08:33 ======================================================================
15:08:33 ERROR: test_all_any_with_dim (test_torch.TestTorch)
15:08:33 ----------------------------------------------------------------------
15:08:33 Traceback (most recent call last):
15:08:33   File "test_torch.py", line 794, in test_all_any_with_dim
15:08:33     [1, 1, 1]]))
15:08:33   File "test_torch.py", line 784, in test
15:08:33     self.assertTrue((r1 == r2).all())
15:08:33 RuntimeError: Expected object of type torch.LongTensor but found type torch.ByteTensor for argument #2 'other'
15:08:33 
15:08:33 ----------------------------------------------------------------------
```

This broken master

**Comment by sighingnow** [comment]
Sorry for the breakage. Thanks for fixing this @soumith .

**Pull request merged**